syft/serde/syft_serializable.py

```diff
@@ -0,0 +1,94 @@
+def get_protobuf_subclasses(cls):
+    """
+        Function to retrieve all classes that implement the protobuf methods from the SyftSerializable class:
+
+        A type that wants to implement the protobuf interface and to use it with syft-proto has
+        to inherit SyftSerializable (directly or from the parent class) and to implement
+        (cannot inherit from parent class):
+            1. bufferize
+            2. unbufferize
+            3. get_protobuf_schema
+
+        If these methods are not implemented, the class won't be enrolled in the types that can use syft-proto.
+    """
+
+    def check_implementation(s):
+        bufferize_implemented = s.bufferize.__qualname__.startswith(s.__name__)
+        unbufferize_implemented = s.unbufferize.__qualname__.startswith(s.__name__)
+        get_protobuf_schema_implemented = s.get_protobuf_schema.__qualname__.startswith(s.__name__)
+        return bufferize_implemented and unbufferize_implemented and get_protobuf_schema_implemented
+
+    original_subclasses = {s for s in cls.__subclasses__() if check_implementation(s)}
+    sub_sets = {
+        s
+        for c in cls.__subclasses__()
+        for s in get_protobuf_subclasses(c)
+        if check_implementation(s)
+    }
+    return original_subclasses.union(sub_sets)
+
+
+def get_msgpack_subclasses(cls):
+    """
+        Function to retrieve all classes that implement the msgpack methods from the SyftSerializable class:
+
+        A type that wants to implement the msgpack interface and to use it in syft has
+        to inherit SyftSerializable (directly or from the parent class) and to implement
+        (cannot inherit from parent class):
+            1. simplify
+            2. detail
+
+        If these methods are not implemented, the class won't be enrolled in the types that can use msgpack.
+    """
+
+    def check_implementation(s):
+        bufferize_implemented = s.simplify.__qualname__.startswith(s.__name__)
+        unbufferize_implemented = s.detail.__qualname__.startswith(s.__name__)
+        return bufferize_implemented and unbufferize_implemented
+
+    original_subclasses = {s for s in cls.__subclasses__() if check_implementation(s)}
+    sub_sets = {
+        s
+        for c in cls.__subclasses__()
+        for s in get_msgpack_subclasses(c)
+        if check_implementation(s)
+    }
+    return original_subclasses.union(sub_sets)
+
+
+class SyftSerializable:
```

Comment:
Do you think this should be an abstract class and its methods abstract methods?

Reply:
Oh, good catch. Yeah, this should never be instantiated directly. What do abstract methods look like in comparison to raising `NotImplementedError`?

Reply:
I think you just have to declare the method with an `abstractmethod` decorator and do nothing in it.
A difference is that an error will be raised as soon as you create an instance of a class inheriting from the abstract one if it doesn't implement all the abstract methods.
At least, that's what I remember from abstract classes in Python ^^

Reply:
That sounds good. 👍

Reply:
(The three of us write some pretty decent Python together.) 😆

Reply:
This will generate some headache on method resolution order (MRO), for example:
SyftSerializable(ABC)
AbstractWorker(ABC, SyftSerializable)

We should drop the ABC from AbstractWorker by doing this (and all the ABCs around the codebase that implements SyftSerializable). Might be a cool idea, not sure if we want it, not sure how it will behave when somethings like this happens, I need to make a few experiments: (here we can't drop Interface2 from the inheritance chain of Interface3)

Interface1(ABC)
Interface2(Interface1)
Interface3(Interface2, SyftSerializable).

Reply:
@karlhigley  any thoughts about this?

Reply:
Let's leave it as is for now. Doesn't seem worth it to incur those headaches.

Reply:
What happens if we switch the order of the parent classes?
```
AbstractWorker(SyftSerializable, ABC)
```
Seems like MRO should work out in that case?